llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -314,20 +315,102 @@ class RegionEndEdges : public ScheduleDAGMutation {
 /// "fixed" SUnits.
 class EmitFixedSUnits : public ScheduleDAGMutation {
 public:
+  struct RegAvailabilityTracker {
+    using fixed_iterator = MachineBasicBlock::iterator;
+    const InstrItineraryData *InstrItins;
+    const TargetRegisterInfo *TRI;
+    std::map<MCRegister, unsigned> RegisterToCycle;
+
+    RegAvailabilityTracker(const InstrItineraryData *Itins,
+                           const TargetRegisterInfo *TRI)
+        : InstrItins(Itins), TRI(TRI) {};
+
+    void updateRegisterAvailability(Register Reg, unsigned Latency) {
+      for (MCRegAliasIterator Ali(Reg, TRI, true); Ali.isValid(); ++Ali) {
+        auto RegCycle = RegisterToCycle.find(*Ali);
+        if (RegCycle == RegisterToCycle.end()) {
+          RegisterToCycle[*Ali] = Latency;
+        } else {
+          RegisterToCycle[*Ali] = std::max(RegCycle->second, Latency);
+        }
+      }
+    }
+
+    // This function calculates the cycles in which each register will be last
+    // materialized, taking into account a specified sequence of timed bundles.
+    // PastTheEndCycles is utilized to project the availability cycle to the
+    // subsequent region.
+    void computeAvailabilityCycles(ArrayRef<AIE::MachineBundle> Bundles,
+                                   bool PastTheEndCycles) {
+      int Cycle = 0;
+      const int TotalCycles = Bundles.size();
+      for (const auto &Bundle : Bundles) {
+        for (MachineInstr *SrcMI : Bundle.getInstrs()) {
+          for (unsigned OpNum = 0; OpNum < SrcMI->getNumOperands(); OpNum++) {
+            unsigned SrcClass = SrcMI->getDesc().getSchedClass();
+            const MachineOperand &MO = SrcMI->getOperand(OpNum);
+            if (!MO.isReg() || !MO.isDef())
+              continue;
+            std::optional<unsigned> OptSrcCycle =
+                InstrItins->getOperandCycle(SrcClass, OpNum);
+            assert(OptSrcCycle);
+            int Latency = *OptSrcCycle;
+            int MaterializationCycle = 0;
+            if (PastTheEndCycles) {
+              int RemainingCycles = TotalCycles - Cycle - 1;
+              MaterializationCycle = std::max(Latency - RemainingCycles - 1, 0);
+            } else {
+              MaterializationCycle = Cycle + Latency;
+            }
+            updateRegisterAvailability(MO.getReg(), MaterializationCycle);
+          }
+        }
+        Cycle++;
+      }
+    }
+
+    unsigned getMaxSrcOperandLatency(const MachineInstr &MI) const {
+      unsigned MaxLatency = 0;
+      for (const MachineOperand &MO : MI.all_uses()) {
+        if (!MO.isReg() || !MO.isUse())
+          continue;
+        for (MCRegAliasIterator Ali(MO.getReg(), TRI, true); Ali.isValid();
```

Comment:
Don't use alias iterator here, because we populated `RegisterToCycle` with all aliases, so we are creating false alias cases, like  `bmh0` aliasing to `bml0`.